Overview function with normalised axis (for cases and deaths)

kirienko opened the pull request:
Here is the attempt of having two Y-axes on the same plot: one is "normal" and another is normalized per 100K people.

There is one thing we didn't think about in advance: in our 14-day incidence rate we _accumulate_ cases, although here we just normalise daily changes. Here what we actually do is we change the units, but the graph preserve its shape. If want to have integral metric (like 14-days), it will be a completely different graph.

The question is: is that type of normalization I introduce in this commit is what we want? If yes, then I can think about more intuitive representations of the two grids.

![image](https://user-images.githubusercontent.com/1487169/91904695-a5953700-eca5-11ea-8e65-f74879b5b411.png)

fangohr commented:
Hi Yury, yes, I was thinking of not changing the data (i.e. no integration / sum over window) and the shape will stay the same.

kirienko commented:
Sorry for the delay with the solution. The reason was mainly our famous PR with the 14-days incidence rate. There we [had broken](https://github.com/oscovida/oscovida/pull/102/files#diff-0a2ce53910eb6d0d4be75a50ccfc3989) some important stuff, namely the `get_population` function. Initially it supported countries and regions within countries, but then it was removed for some good reasons. And since some parts were not covered with tests, we didn't noticed that. (My bad, I personally gave an LGTM for that.)

Here I fixed some of those issues (and probably introduces some new ones :smile:):
* I restored initial `get_population` function under the name `population`. Unfortunately it only supports for Germany (i.e. _Bundesländer_) but not, say, states within the US;
* I fixed broken `normalise` feature for `make_compare_plot`;
* [by-product] I also tried to decouple the fetching data and the label creation (see `get_region_label`) — I fully agree that it'd be better to do a separate PR for that, but I decided that I've already delayed for too long, and that's why I decided not to spend additional time for separating that piece of code;
* the tests were updated accordingly.

kirienko commented:
Damn, the test for `make_compare_plot_hungary()` is failing now. Have no idea why, but it's not me!

fangohr commented:
Proposed fine tuning of plots (from discussion):

- put relative incidence on left y-axis, leave matplotlib to choose sensible labels
- put total numbers on the right y-axis
- use the gridlines from left axis for right axis, by using non-round numbers on the right

fangohr commented:
- double check plot on title page (https://oscovida.github.io/staging/111/index.html): overlap of y-axis label on the right with numbers

ebadkamil commented on oscovida/oscovida.py:
```diff
@@ -478,6 +478,39 @@ def get_population() -> pd.DataFrame:
+        if country in df.index and region is None:
+            return int(df.loc[country].population)
+        else:
+            return None
```
`return`

ebadkamil commented on oscovida/oscovida.py:
```diff
@@ -749,28 +783,51 @@ def plot_daily_change(ax, series, color, labels=None):
     if labels is None:
-        label = ""
-        region = ""
+        type_label = ""
```
```python
type_label = ""
region_label = ""
if labels:
    region_label, type_label = labels



```

ebadkamil commented on oscovida/oscovida.py:
```diff
@@ -749,28 +783,51 @@ def plot_daily_change(ax, series, color, labels=None):
     else:
-        region, label = labels
+        region_label, type_label = labels
+    ax_label = region_label + " new " + type_label
```
`ax_label = f"{region_label} new {type_label}"`

ebadkamil commented on oscovida/oscovida.py:
```diff
@@ -749,28 +783,51 @@ def plot_daily_change(ax, series, color, labels=None):
+        habitants = population(country=country, region=region)
+    else:
+        habitants = population(region_label)
+    if region_label is not "" and habitants:
```
`if region_label and habitants:`

ebadkamil commented on oscovida/oscovida.py:
```diff
@@ -1107,8 +1164,43 @@ def plot_reproduction_number(ax, series, color_g='C1', color_R='C4',
+    _country = country.casefold()
+    region_label = None
+    if _country == 'germany':
+        # if region == None and subregion == None:
```
remove commented code

ebadkamil commented on oscovida/oscovida.py:
```diff
@@ -1107,8 +1164,43 @@ def plot_reproduction_number(ax, series, color_g='C1', color_R='C4',
+            region_label = f"Germany-{region}"
+        elif region is None and subregion is not None:
+            region_label = f"Germany-{subregion}"
+        elif region is None and subregion is not None:
```
repeated as above

ebadkamil commented on oscovida/oscovida.py:
```diff
@@ -1107,8 +1164,43 @@ def plot_reproduction_number(ax, series, color_g='C1', color_R='C4',
+        elif region is None and subregion is not None:
+            region_label = f"Germany-{subregion}"
+
+    elif _country.lower() == 'us' and region != None:
```
`elif _country.lower() == 'us' and region is not None:`

ebadkamil commented on oscovida/oscovida.py:
```diff
@@ -1577,13 +1652,24 @@ def has_twin(ax: plt.Axes) -> bool:
+    # obviously, f(left_min) -> right_min and f(left_max) -> right_max, and points in between are mapped proportionally
+    left = ax_left.get_ylim()
+    right = ax_right.get_ylim()
+    f = lambda x: right[0] + (x - left[0]) / (left[1] - left[0]) * (right[1] - right[0])
```
`return right[0] + (ax_left.get_yticks() - left[0]) / (left[1] - left[0]) * (right[1] - right[0])`

ebadkamil commented on tests/test_corona.py:
```diff
@@ -81,12 +81,11 @@ def test_fetch_data_hungary():
     # Hungary related
     with_local = c.choose_random_counties(exclude_region="Baranya", size=18)
-    print(with_local)
+    # print(with_local)
```
remove commented code

ebadkamil reviewed (commented):
besides minor comments LGTM.

ebadkamil commented:
also commits can be squashed.

ebadkamil commented on oscovida/oscovida.py:
```diff
@@ -736,7 +769,8 @@ def compute_daily_change(series):
 
 
-def plot_daily_change(ax, series, color, labels=None):
+def plot_daily_change(ax, series: pd.Series, color: str, labels: Tuple[str, str] = None,
```
```python
def plot_daily_change(ax: plt.Axes, series: pd.Series, color: str, labels: Tuple[str, str] = None,
                                    country: str = None, region: str = None, subregion: str = None):
```
